commands/instances.go

```diff
@@ -173,120 +160,142 @@ func Create(req *rpc.CreateRequest, extraUserAgent ...string) (*rpc.CreateRespon
 // Failures don't stop the loading process, in case of loading failure the Platform or library
 // is simply skipped and an error gRPC status is sent to responseCallback.
 func Init(req *rpc.InitRequest, responseCallback func(r *rpc.InitResponse)) error {
-	if responseCallback == nil {
-		responseCallback = func(r *rpc.InitResponse) {}
-	}
 	instance := instances[req.Instance.Id]
 	if instance == nil {
 		return &arduino.InvalidInstanceError{}
 	}
 
+	// Setup callback functions
+	if responseCallback == nil {
+		responseCallback = func(r *rpc.InitResponse) {}
+	}
+	responseError := func(st *status.Status) {
+		responseCallback(&rpc.InitResponse{
+			Message: &rpc.InitResponse_Error{
+				Error: st.Proto(),
+			},
+		})
+	}
+	taskCallback := func(msg *rpc.TaskProgress) {
+		responseCallback(&rpc.InitResponse{
+			Message: &rpc.InitResponse_InitProgress{
+				InitProgress: &rpc.InitResponse_Progress{
+					TaskProgress: msg,
+				},
+			},
+		})
+	}
+	downloadCallback := func(msg *rpc.DownloadProgress) {
+		responseCallback(&rpc.InitResponse{
+			Message: &rpc.InitResponse_InitProgress{
+				InitProgress: &rpc.InitResponse_Progress{
+					DownloadProgress: msg,
+				},
+			},
+		})
+	}
+
 	// We need to clear the PackageManager currently in use by this instance
 	// in case this is not the first Init on this instances, that might happen
 	// after reinitializing an instance after installing or uninstalling a core.
 	// If this is not done the information of the uninstall core is kept in memory,
 	// even if it should not.
-	instance.PackageManager.Clear()
+	pm := instance.PackageManager
+	pm.Clear()
+
+	// Try to extract profile if specified
+	var profile *sketch.Profile
+	if req.GetProfile() != "" {
+		sk, err := sketch.New(paths.New(req.GetSketchPath()))
+		if err != nil {
+			return &arduino.InvalidArgumentError{Cause: err}
+		}
+		profile = sk.GetProfile(req.GetProfile())
```

Comment:
A `nil` return value must be handled here, otherwise you will get a panic at line 218 if the specified profile does not exist.